### PR TITLE
implement server-side on() events

### PR DESCRIPTION
**Pull request by romaninsh**
This pull requests aims to make this code work:

``` php
$b = new Button();
$b->on('click', function($js){ $js->text(rand(1,20)); });
```

Clicking a button here would execute request to the server that would generate the new label for a button.

### CODE REVIEW

**Comment by romaninsh** [comment]
I am OK with the remaining 4 code-climate issues. I don't want to disable them because we will need to address them at some point.

File: docs/js.rst [hunk 1]
```diff
@@ -72,75 +193,149 @@ will output:
...
-    Some of the clases that implement jsExpressionable are:
-     
-    - jsExpression
-    - jsChain
-    - View
+    $b1 = $buttons->add(new Button('One'));
+    $b2 = $buttons->add(new Button('Two'));
+    $b3 = $buttons->add(new Button('Three'));
 
+    $buttons->on('click', '.button', $b3->js()->hide());
```
Comment (DarkSide666) [inline docs/js.rst]: Wrong example!

File: src/View.php [hunk 2]
```diff
@@ -478,7 +495,7 @@ public function getHTML()
      *
```
Comment (DarkSide666) [inline src/View.php]: no description what $extra parameter is for

File: src/View.php [hunk 3]
```diff
@@ -555,8 +576,51 @@ public function on($event, $selector = null, $action = null)
...
+                    } else {
+                        throw new Exception(['Incorrect callback. Must be string or action.', 'r'=>$r]);
+                    }
+                }
+
+                $ajaxec = implode(";\n", array_map(function (jQuery $r) {
+                    return $r->jsRender();
+                }, $actions));
+
+                echo json_encode(['success'=>true, 'message'=>'Hello World', 'eval'=>$ajaxec]);
```
Comment (DarkSide666) [inline src/View.php]: Hello world???

File: src/View.php [hunk 4]
```diff
@@ -555,8 +576,51 @@ public function on($event, $selector = null, $action = null)
...
+                exit;
+            });
+
+            $thisAction->api(['on'=>'now', 'url'=>$cb->getURL(), 'obj'=>new jsExpression('this')]);
```
Comment (DarkSide666) [inline src/View.php]: $thisAction->api ? what is that?

File: src/jsExpression.php [hunk 5]
```diff
@@ -73,4 +73,94 @@ function ($matches) use (&$namelessCount) {
 
         return trim($res);
     }
+
+    /**
+     * Provides replacement for json_encode that will respect jsExpressionable objects
+     * and call jsRender() for them instead of escaping.
+     *
+     * @param mixed @arg anything
```
Comment (DarkSide666) [inline src/jsExpression.php]: should be $arg instead of @arg

File: src/jsAPI.php [hunk 6]
```diff
@@ -0,0 +1,19 @@
+<?php
```
Comment (DarkSide666) [inline src/jsAPI.php]: what is this class for? where is it used?

Comment (romaninsh) [inline src/jsAPI.php]: removing it, it's confusing.

File: src/jsExpression.php (hunk 5 above)
Comment (romaninsh) [inline src/jsExpression.php]: fixed

File: src/View.php (hunk 4 above)
Comment (romaninsh) [inline src/View.php]: that's Semantic UI ;) 

http://semantic-ui.com/behaviors/api.html

File: src/View.php (hunk 3 above)
Comment (romaninsh) [inline src/View.php]: This is changed in the next PR, so i'll leave it to avoid conflicts.

File: src/View.php (hunk 2 above)
Comment (romaninsh) [inline src/View.php]: it's changed in the next PR, so leaving.

File: docs/js.rst (hunk 1 above)
Comment (romaninsh) [inline docs/js.rst]: There is major overhaul on js.rst in the next PR.

**Pull request merged**